frontend/udp/parser.go

```diff
@@ -99,41 +103,69 @@ func ParseAnnounce(r Request, allowIPSpoofing bool) (*bittorrent.AnnounceRequest
 
 // handleOptionalParameters parses the optional parameters as described in BEP
 // 41 and updates an announce with the values parsed.
-func handleOptionalParameters(packet []byte) (params bittorrent.Params, err error) {
-	if len(packet) <= 98 {
-		return
+func handleOptionalParameters(packet []byte) (bittorrent.Params, error) {
+	if len(packet) == 0 {
+		return emptyURLData, nil
 	}
 
-	optionStartIndex := 98
-	for optionStartIndex < len(packet)-1 {
-		option := packet[optionStartIndex]
+	var buf bytes.Buffer
+
+	for i := 0; i < len(packet); {
+		option := packet[i]
 		switch option {
 		case optionEndOfOptions:
-			return
-
+			return parseURLData(buf.String())
 		case optionNOP:
-			optionStartIndex++
-
+			i++
 		case optionURLData:
-			if optionStartIndex+1 > len(packet)-1 {
-				return params, errMalformedPacket
+			if i+1 >= len(packet) {
+				return nil, errMalformedPacket
 			}
 
-			length := int(packet[optionStartIndex+1])
-			if optionStartIndex+1+length > len(packet)-1 {
-				return params, errMalformedPacket
+			length := int(packet[i+1])
+			if i+2+length > len(packet) {
+				return nil, errMalformedPacket
 			}
 
-			// TODO(chihaya): Actually parse the URL Data as described in BEP 41
-			// into something that fulfills the bittorrent.Params interface.
+			n, err := buf.Write(packet[i+2 : i+2+length])
```

Comment:
To concatenate multiple URLData blocks. Then again I don't know how often you'd have multiple blocks of URLData anyways, so using them as strings and concatenating in the improbable case might be the better choice.

We could pool the buffers?